File: web/frontend/src/api/agents.ts
import { launcherFetch } from "@/api/http"

export interface Agent {
  slug: string
  name: string
  description: string
  system_prompt: string
  model: string
  tool_permissions: string[]
  status: "enabled" | "disabled"
  created_at: number
  updated_at: number
}

export interface AgentListResponse {
  agents: Agent[]
}

export interface AgentCreateRequest {
  name: string
  description?: string
  system_prompt: string
  model: string
  tool_permissions?: string[]
}

export interface AgentUpdateRequest {
  name?: string
  description?: string
  system_prompt?: string
  model?: string
  tool_permissions?: string[]
  status?: string
}

export async function listAgents(): Promise<AgentListResponse> {
  const res = await launcherFetch("/api/agents")
  if (!res.ok) throw new Error(`Failed to list agents: ${res.status}`)
  return res.json()
}

export async function getAgent(slug: string): Promise<Agent> {
  const params = new URLSearchParams({ slug })
  const res = await launcherFetch(`/api/agent?${params.toString()}`)
  if (!res.ok) throw new Error(`Failed to get agent: ${res.status}`)
  return (await res.json()).agent
}

export async function createAgent(data: AgentCreateRequest): Promise<Agent> {
  const res = await launcherFetch("/api/agent/create", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  })
  if (!res.ok) throw new Error(`Failed to create agent: ${res.status}`)
  return (await res.json()).agent
}

export async function updateAgent(slug: string, data: AgentUpdateRequest): Promise<Agent> {
  const params = new URLSearchParams({ slug })
  const res = await launcherFetch(`/api/agent/update?${params.toString()}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  })
  if (!res.ok) throw new Error(`Failed to update agent: ${res.status}`)
  return (await res.json()).agent
}

export async function deleteAgent(slug: string): Promise<{ status: string }> {
  const params = new URLSearchParams({ slug })
  const res = await launcherFetch(`/api/agent/delete?${params.toString()}`, {
    method: "DELETE",
  })
  if (!res.ok) throw new Error(`Failed to delete agent: ${res.status}`)
  return res.json()
}

export async function importAgent(content: string): Promise<Agent> {
  const res = await launcherFetch("/api/agent/import", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
  })
  if (!res.ok) throw new Error(`Failed to import agent: ${res.status}`)
  return (await res.json()).agent
}

File: web/backend/api/agents.go
package api

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	manager "github.com/sipeed/picoclaw/pkg/agent/manager"
)

// Type aliases to use manager types directly
type agent = manager.Agent
type agentListResponse = manager.AgentListResponse
type agentResponse struct {
	Agent *manager.Agent `json:"agent,omitempty"`
}

// agentCreateRequest uses same fields as manager.AgentCreateRequest
type agentCreateRequest = manager.AgentCreateRequest

// agentUpdateRequest uses same fields as manager.AgentUpdateRequest
type agentUpdateRequest = manager.AgentUpdateRequest

func (h *Handler) registerAgentRoutes(mux *http.ServeMux) {
	mux.HandleFunc("/api/agents", h.handleListAgents)
	mux.HandleFunc("/api/agent", h.handleGetAgent)
	mux.HandleFunc("/api/agent/create", h.handleCreateAgent)
	mux.HandleFunc("/api/agent/update", h.handleUpdateAgent)
	mux.HandleFunc("/api/agent/delete", h.handleDeleteAgent)
	mux.HandleFunc("/api/agent/import", h.handleImportAgent)
}

func (h *Handler) handleListAgents(w http.ResponseWriter, r *http.Request) {
	mgr := manager.NewManager("")
	agents, err := mgr.ListAgents()
	if err != nil {
		http.Error(w, fmt.Sprintf("Failed to list agents: %v", err), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(manager.AgentListResponse{Agents: agents})
}

func (h *Handler) handleGetAgent(w http.ResponseWriter, r *http.Request) {
	slug := strings.TrimSpace(r.URL.Query().Get("slug"))
	if slug == "" {
		http.Error(w, "slug is required", http.StatusBadRequest)
		return
	}

	mgr := manager.NewManager("")
	a, err := mgr.GetAgent(slug)
	if err != nil {
		http.Error(w, fmt.Sprintf("Agent not found: %s", slug), http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(agentResponse{Agent: a})
}

func (h *Handler) handleCreateAgent(w http.ResponseWriter, r *http.Request) {
	var req manager.AgentCreateRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, fmt.Sprintf("Invalid JSON: %v", err), http.StatusBadRequest)
		return
	}
	if req.Name == "" {
		http.Error(w, "name is required", http.StatusBadRequest)
		return
	}
	if req.SystemPrompt == "" {
		http.Error(w, "system_prompt is required", http.StatusBadRequest)
		return
	}
	if req.Model == "" {
		http.Error(w, "model is required", http.StatusBadRequest)
		return
	}
	mgr := manager.NewManager("")
	created, err := mgr.CreateAgent(req)
	if err != nil {
		http.Error(w, fmt.Sprintf("Failed to create agent: %v", err), http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(agentResponse{Agent: created})
}

func (h *Handler) handleUpdateAgent(w http.ResponseWriter, r *http.Request) {
	slug := strings.TrimSpace(r.URL.Query().Get("slug"))
	if slug == "" {
		http.Error(w, "slug is required", http.StatusBadRequest)
		return
	}

	var req manager.AgentUpdateRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, fmt.Sprintf("Invalid JSON: %v", err), http.StatusBadRequest)
		return
	}

	mgr := manager.NewManager("")
	updated, err := mgr.UpdateAgent(slug, req)
	if err != nil {
		http.Error(w, fmt.Sprintf("Failed to update agent: %v", err), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(agentResponse{Agent: updated})
}

func (h *Handler) handleDeleteAgent(w http.ResponseWriter, r *http.Request) {
	slug := strings.TrimSpace(r.URL.Query().Get("slug"))
	if slug == "" {
		http.Error(w, "slug is required", http.StatusBadRequest)
		return
	}

	mgr := manager.NewManager("")
	if err := mgr.DeleteAgent(slug); err != nil {
		http.Error(w, fmt.Sprintf("Failed to delete agent: %v", err), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{"status": "ok"})
}

type agentImportRequest struct {
	Content string `json:"content" binding:"required"`
}

func (h *Handler) handleImportAgent(w http.ResponseWriter, r *http.Request) {
	var req agentImportRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, fmt.Sprintf("Invalid JSON: %v", err), http.StatusBadRequest)
		return
	}

	if req.Content == "" {
		http.Error(w, "content is required", http.StatusBadRequest)
		return
	}

	mgr := manager.NewManager("")
	created, err := mgr.ImportAgent(req.Content)
	if err != nil {
		http.Error(w, fmt.Sprintf("Failed to import agent: %v", err), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(agentResponse{Agent: created})
}